Softeq.XToolkit.WhiteLabel.iOS/AppDelegateBase.cs

```diff
@@ -27,7 +27,7 @@ public override bool FinishedLaunching(UIApplication application, NSDictionary l
             _rootViewController = CreateRootViewController();
 
             InitializeMainWindow(_rootViewController);
-            InitializeWhiteLabelRuntime();
+            InitializeWhiteLabelRuntime(application, launchOptions);
 
```

Comment:
What about adding an additional method below instead?

Reply:
You mean additional `InitializeWhiteLabelRuntime()` overload? I'm not sure it is possible, as in the end we need to pass `launchOptions` into bootstrapper. In current implementation it is not possible to share bootstrapper between two methods

Reply:
I guess as an option we can extract Bootstrapper initialization into a separate method.. but then it won't be logically a part of "White Label initialization"

Reply:
In general,  I'm not sure that we need these signature changes, what is the reason? What about creating `PushNotificationsAppDelegate` for example?

Reply:
Well, it's a part of platform-specific class, `AppDelegateBase`, so it should be able to use platform-specific API.
The reason behind this change is that one of our third-party integrations require `launchOptions` for initialization, and it would be preferable not to use this integration directly (especially providing that it's a singleton) but cover it behind interface and use DI.
Without this change, the only option is to cache `launchOptions` in the `AppDelegate.FinishedLaunching()` BEFORE calling base implementation (because by the end of that call Bootstrapper would be already created), so they can be used in `CreateBootstrapper()` - this actually looks like a hack:

```
private NSDictionary _launchOptions;

public override bool FinishedLaunching(UIApplication application, NSDictionary launchOptions)
{
      _launchOptions = launchOptions;
          
      var result = base.FinishedLaunching(application, launchOptions);

      ...

      return result;
}

protected override IBootstrapper CreateBootstrapper()
{
    return new Bootstrapper(_launchOptions);
}
```

Reply:
Is it really required to change these methods and transfer two parameters through all these classes and methods to simplify init of a single 3rd-party SDK?

Reply:
Well, yes, it is really required, as you can see above (that, of course, if we encourage users of our toolkit to use DI and our implementation of Bootstrapper).

I guess you meant to ask "Is it worth it?" Well, also yes. 
It's a single 3rd-party SDK now, but it doesn't mean that none of the 3rd-party integrations won't need it ever again. Besides, I didn't really drag those parameters through any additional classes (they are already a part of `AppDelegateBase.FinishedLaunching(UIApplication application, NSDictionary launchOptions)`, we just have been ignoring them before), and only through one additional method before reaching the bootstrapper (`InitializeWhiteLabelRuntime()`) - so as a solution we might extract bootstrapper creation from WhiteLabel initialization, like this:
```
public override bool FinishedLaunching(UIApplication application, NSDictionary launchOptions)
{
    _rootViewController = CreateRootViewController();

    InitializeMainWindow(_rootViewController);
    
    var bootstrapper = CreateBootstrapper(application, launchOptions);
    InitializeWhiteLabelRuntime(bootstrapper);

    return true;
}
```

Reply:
In general, I would like to keep methods simple. Currently, `InitializeWhiteLabelRuntime` method for Android and iOS looks nearly the same and I prefer to not add unnecessary platfrom-specific arguments to it. `CreateBootstrapper` registers dependencies and in common cases doesn't require platfrom-specific arguments, otherwise, this method will be used for getting options for whatever not only deps.

I suggest something like this:

- simple change
- no breaking change
- flexibility in use from any place of child delegate

```cs
public abstract class AppDelegateBase // ...
{
	protected NSDictionary LaunchOptions { get; private set; }

	public override bool FinishedLaunching(UIApplication application, NSDictionary launchOptions)
	{
	    LaunchOptions = launchOptions;
		//...
	}
}
```
```cs
//...
protected override IBootstrapper CreateBootstrapper()
{
    return new CustomIosBootstrapper(LaunchOptions);
}
//...
```

now it can be used wherever needed: `CreateBootstrapper`, `InitializeNavigation`, etc.
What do you think?

Reply:
OK, I guess it would do the trick. But it seems like this decision is forced by "no breaking change" rule, and at the same time this whole PR is a huge breaking change.

Reply:
Looks so, but this PR is only about PushNotifications breaking changes

Reply:
And also it seems like by extending the scope of the launchOptions by default we would introduce a memory leak for every user who won't use this property outside of the `FinishedLaunching()` method

Reply:
Not a memory leak it's just one small dictionary in memory.

Reply:
It's not a big one, yes, but it is a memory leak by [definition](https://en.wikipedia.org/wiki/Memory_leak):

> In computer science, a memory leak is a type of resource leak that occurs when a computer program incorrectly manages memory allocations in a way that **memory which is no longer needed is not released**.

Reply:
formal yes, but leak it's a **process** of decrease/degradation, usually uncontrolled
in our case, it's just a single drop :)